pkg/manager/member/pvc_cleaner.go

```diff
@@ -40,37 +49,161 @@ type PVCCleanerInterface interface {
 	podLister corelisters.PodLister,
 	pvcControl controller.PVCControlInterface,
-	pvcLister corelisters.PersistentVolumeClaimLister) PVCCleanerInterface {
+	pvcLister corelisters.PersistentVolumeClaimLister,
+	pvLister corelisters.PersistentVolumeLister,
+	pvControl controller.PVControlInterface) PVCCleanerInterface {
 	return &realPVCCleaner{
+		kubeCli,
 		podLister,
 		pvcControl,
 		pvcLister,
+		pvLister,
+		pvControl,
 	}
 }
 
 func (rpc *realPVCCleaner) Clean(tc *v1alpha1.TidbCluster) (map[string]string, error) {
+	if skipReason, err := rpc.cleanScheduleLock(tc); err != nil {
+		return skipReason, err
+	}
+
+	if !tc.Spec.EnablePVReclaim {
+		// disable PV reclaim, return directly.
+		return nil, nil
+	}
+	return rpc.reclaimPV(tc)
+}
+
+func (rpc *realPVCCleaner) reclaimPV(tc *v1alpha1.TidbCluster) (map[string]string, error) {
 	ns := tc.GetNamespace()
 	tcName := tc.GetName()
+
 	// for unit test
 	skipReason := map[string]string{}
 
-	selector, err := label.New().Instance(tc.GetLabels()[label.InstanceLabelKey]).Selector()
+	pvcs, err := rpc.listAllPVCs(tc)
 	if err != nil {
-		return skipReason, fmt.Errorf("cluster %s/%s assemble label selector failed, err: %v", ns, tcName, err)
+		return skipReason, err
 	}
 
-	pvcs, err := rpc.pvcLister.PersistentVolumeClaims(ns).List(selector)
+	for _, pvc := range pvcs {
+		pvcName := pvc.GetName()
+		l := label.Label(pvc.Labels)
+		if !(l.IsPD() || l.IsTiKV()) {
+			skipReason[pvcName] = skipReasonPVCCleanerIsNotPDOrTiKV
+			continue
+		}
+
+		if pvc.Status.Phase != corev1.ClaimBound {
+			// If pvc is not bound yet, it will not be processed
+			skipReason[pvcName] = skipReasonPVCCleanerPVCNotBound
+			continue
+		}
+
+		if pvc.DeletionTimestamp != nil {
+			// PVC has been deleted, skip it
+			skipReason[pvcName] = skipReasonPVCCleanerPVCHasBeenDeleted
+			continue
+		}
+
+		if len(pvc.Annotations[label.AnnPVCDeferDeleting]) == 0 {
+			// This pvc has not been marked as defer delete PVC, can't reclaim the PV bound to this PVC
+			skipReason[pvcName] = skipReasonPVCCleanerIsNotDeferDeletePVC
+			continue
+		}
+
+		// PVC has been marked as defer delete PVC, try to reclaim the PV bound to this PVC
+		podName, exist := pvc.Annotations[label.AnnPodNameKey]
+		if !exist {
+			// PVC has not pod name annotation, this is an unexpected PVC, skip it
+			skipReason[pvcName] = skipReasonPVCCleanerPVCNotHasPodNameAnn
+			continue
+		}
+
+		_, err := rpc.podLister.Pods(ns).Get(podName)
+		if err == nil {
+			// PVC is still referenced by this pod, can't reclaim PV
+			skipReason[pvcName] = skipReasonPVCCleanerPVCeferencedByPod
+			continue
+		}
+		if !errors.IsNotFound(err) {
+			return skipReason, fmt.Errorf("cluster %s/%s get pvc %s pod %s from local cache failed, err: %v", ns, tcName, pvcName, podName, err)
+		}
+
+		// if pod not found in cache, re-check from apiserver directly to make sure the pod really not exist
+		_, err = rpc.kubeCli.CoreV1().Pods(ns).Get(podName, metav1.GetOptions{})
```

Comment:
Why not find Pod from apiServer directly? It seems we would alway to chek whether pod exist or not from apiServer.

Reply:
The object may have been modified during the first query to the second query, so when you need to make some decisions based on this object, we need to query the object from apiserver again to confirm this object has not been modified during this time, as much as possible to ensure that our decision is accurate.

Reply:
> Why not find Pod from apiServer directly?
This avoids API request as possible as we can because we don't need to clean the PVC in time, so we can skip if the pod exist in the cache (even if the pod may do not exist in the apiserver). However, if the pod does not exist in the cache, we must re-check the apiserver before deleting the PVC because a new pod may be just created by the statefulset controller to use this PVC.